packages/core/src/Cardano/Address/Address.ts

```diff
@@ -0,0 +1,739 @@
+   */
+  asReward(): RewardAddress | undefined {
+    return RewardAddress.fromAddress(this);
+  }
+
+  /**
+   * Gets this Address instance as a PointerAddress (undefined if Address is not a valid PointerAddress).
+   */
+  asPointer(): PointerAddress | undefined {
+    return PointerAddress.fromAddress(this);
+  }
+
+  /**
+   * Gets this Address instance as a EnterpriseAddress (undefined if Address is not a valid EnterpriseAddress).
+   */
+  asEnterprise(): EnterpriseAddress | undefined {
+    return EnterpriseAddress.fromAddress(this);
+  }
+
+  /**
+   * Gets this Address instance as a BaseAddress (undefined if Address is not a valid BaseAddress).
+   */
+  asBase(): BaseAddress | undefined {
+    return BaseAddress.fromAddress(this);
+  }
+
+  /**
+   * Gets the address type.
+   */
+  getType(): AddressType {
+    return this.#props.type;
+  }
+
+  /**
+   * Gets the address network id.
+   */
+  getNetworkId(): NetworkId {
+    if (this.#props.type === AddressType.Byron) {
+      // Byron addresses do not encode the network identifier, however, they include an optional attribute
+      // 'network magic', the network magic is only set to specify which testnet this address is for (if any). We can
+      // then assume that if the network magic is undefined the network id is mainnet; otherwise; testnet.
+      if (this.#props.byronAddressContent?.attrs.magic === undefined) return NetworkId.Mainnet;
+
+      return NetworkId.Testnet;
+    }
+
+    return this.#props.networkId!;
+  }
+
+  /**
+   * Gets the address properties.
+   */
+  getProps(): AddressProps {
+    return this.#props;
+  }
+
+  /**
+   * Gets the address bech32 prefix.
+   */
+  // eslint-disable-next-line complexity
+  private static getBech32Prefix(type: AddressType, networkId: NetworkId): string {
+    let prefix = '';
+    switch (type) {
+      case AddressType.BasePaymentKeyStakeKey:
+      case AddressType.BasePaymentScriptStakeKey:
+      case AddressType.BasePaymentKeyStakeScript:
+      case AddressType.BasePaymentScriptStakeScript:
+      case AddressType.PointerKey:
+      case AddressType.PointerScript:
+      case AddressType.EnterpriseKey:
+      case AddressType.EnterpriseScript:
+        prefix = 'addr';
+        break;
+      case AddressType.RewardKey:
+      case AddressType.RewardScript: {
+        prefix = 'stake';
+        break;
+      }
+      default:
+        throw new Error('Invalid address'); // Shouldn't happen
+    }
+
+    prefix += networkId === 0 ? '_test' : '';
+
+    return prefix;
+  }
+
+  /**
+   * There are currently 4 types of Shelley Base addresses, summarized below:
+   *
+   * - 0000 PaymentKeyHash StakeKeyHash
+   * - 0001 ScriptHash     StakeKeyHash
+   * - 0010 PaymentKeyHash ScriptHash
+   * - 0011 ScriptHash     ScriptHash
+   *
+   * @param type The address type.
+   * @param data The serialized address data.
+   */
+  private static unpackBaseAddressParts(type: number, data: Uint8Array): Address {
```

Comment:
Better to move these to their respective classes, then you can generalise to `packParts` and `unpackParts`. It would reduce the complexity of this file too (still over 700 lines)